File: mstlo-python/examples/batch_update.py
import mstlo_python as mstlo

# Create a simple monitor for batch processing
phi_batch = mstlo.parse_formula("x > 10.0")
batch_monitor = mstlo.Monitor(phi_batch, semantics="Rosi")

# Signal-major form: a dict mapping each signal name to its (value, timestamp)
# samples. Use this when you have one trace per signal.
batch_steps = {
    "x": [
        (5.0, 0.0),  # x=5 at t=0 (robustness: 5-10 = -5)
        (15.0, 1.0),  # x=15 at t=1 (robustness: 15-10 = 5)
        (8.0, 2.0),  # x=8 at t=2 (robustness: 8-10 = -2)
        (12.0, 3.0),  # x=12 at t=3 (robustness: 12-10 = 2)
    ]
}

# Process all steps at once
output = batch_monitor.update_batch(batch_steps)

print("Batch Update Results:")
print(output)

# Flat form: an iterable of (signal, value, timestamp) tuples. Use this when
# samples from different signals are interleaved, e.g. read from a log.
interleaved_monitor = mstlo.Monitor(phi_batch, semantics="Rosi")
output = interleaved_monitor.update_batch(
    [
        ("x", 5.0, 0.0),
        ("x", 15.0, 1.0),
        ("x", 8.0, 2.0),
        ("x", 12.0, 3.0),
    ]
)

print("\nSame trace, flat form:")
print(output)

File: mstlo/examples/batch_update.rs
//! Feed a whole trace to a monitor at once with [`StlMonitor::update_batch`].
//!
//! Run with: `cargo run --example batch_update`

use mstlo::monitor::*;
use mstlo::{steps, stl};

fn main() {
    let formula = stl!(x > 10.0);

    let mut batch_monitor = StlMonitor::builder()
        .formula(formula.clone())
        .semantics(Rosi)
        .build()
        .expect("Failed to build monitor");

    // Signal-major form: one trace per signal, so the name is written once.
    // Use this when you have the samples grouped by signal.
    let output = batch_monitor.update_batch(&steps! {
        "x": [
            (5.0, 0s),  // robustness: 5 - 10 = -5
            (15.0, 1s), // robustness: 15 - 10 = 5
            (8.0, 2s),  // robustness: 8 - 10 = -2
            (12.0, 3s), // robustness: 12 - 10 = 2
        ],
    });

    println!("Batch Update Results:");
    println!("{output}");

    // Flat form: each entry is a `step!` argument list. Use this when samples
    // from different signals are interleaved, e.g. read from a log.
    let mut interleaved_monitor = StlMonitor::builder()
        .formula(formula)
        .semantics(Rosi)
        .build()
        .expect("Failed to build monitor");

    let output = interleaved_monitor.update_batch(&steps![
        ("x", 5.0, 0s),
        ("x", 15.0, 1s),
        ("x", 8.0, 2s),
        ("x", 12.0, 3s),
    ]);

    println!("\nSame trace, flat form:");
    println!("{output}");
}
